### PR TITLE
[Core] Activate windows colors in terminal

### PR DESCRIPTION
**Pull request by loumalouomega**
This PR activates the colors for the Windows terminal. An example with the [ConEmu](https://conemu.github.io/) (cmd.exe)

image

Can anyone tried it with VS?, I usedMinGW

### CODE REVIEW

**Review by philbucher — CHANGES_REQUESTED** [review]
As much as I would love this but we need sth to disable it. The problem is if you redirect the stream to sth that does not support colors (like a file or cluster-OSs) then the color codes are printed and this really messes up the output.

I have implemented a mechanism for this in the CoSimApp, but only in python and I am not sure if it is easily portable to cpp

Related to #5643

**Comment by loumalouomega** [comment]
Additionally it fails in appveyor :P

**Comment by loumalouomega** [comment]
> 
> 
> As much as I would love this but we need sth to disable it. The problem is if you redirect the stream to sth that does not support colors (like a file or cluster-OSs) then the color codes are printed and this really messes up the output.
> 
> I have implemented a mechanism for this in the CoSimApp, but only in python and I am not sure if it is easily portable to cpp
> 
> Related to #5643

We can add a flag in the compilation like -D_USE_ANSI_OUTPUT, so we can deactivate (deactivate in running time in painful, I did it in the contact app long time ago and it is painful)

**Comment by philbucher** [comment]
that is what I suspected.
I would be fine with having a flag to activate the colors
But why `USE_ANSI_OUTPUT`? why not `ENABLE_COLORED_OUTPUT` or sth like this?

**Comment by philbucher** [comment]
the appveyor error is similar to the one in #5465 , no idea what those mean...
@roigcarlo :)

File: kratos/utilities/color_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// External includes
+
+namespace Kratos
+{
+namespace ColorUtilities
+{
+void InitializeSTDOutput()
+{
+#ifdef KRATOS_COMPILED_IN_WINDOWS
+    SetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), ENABLE_VIRTUAL_TERMINAL_PROCESSING);
```
Comment (philbucher) [inline kratos/utilities/color_utilities.cpp:37]: why is this needed?

If this is needed, then I think we should integrate it globally, i.e. when launching the Kernel and not each time you need it (as you do now)

Or am I overseeing sth?

Comment (loumalouomega) [reply kratos/utilities/color_utilities.cpp:37]: I will explain. windows.h is a pain in the ass, it conflict with everything, it is the conflict [lich](https://www.youtube.com/watch?v=FrlymHW0qU8), the only way to content its conflict power is to confine it in a cpp, so it is harmful. This was the only way I can ensure you, if you include it in a header all flags will fail (for some Microsoft  reason). The others ones are for compatibility reasons, specially for MinGW

**Comment by loumalouomega** [comment]
@philbucher now the color is not mandatory

**Comment by loumalouomega** [comment]
BTW, in my computer compilation works

File: kratos/includes/table_stream.h [hunk 2]
```diff
@@ -84,6 +85,9 @@ class TableStream
             mSeparator(Separator),
             mBoldFont(UseBoldFont)
     {
+        // Initialize STD output
+        ColorUtilities::InitializeSTDOutput();
```
Comment (philbucher) [inline kratos/includes/table_stream.h:89]: overall I am fine with this PR but this function confuses me, esp how / when it should be used

Comment (philbucher) [reply kratos/includes/table_stream.h:89]: @roigcarlo could you please also take a look?

Comment (loumalouomega) [reply kratos/includes/table_stream.h:89]: Basically always youy need bold/colors, In Unix does nothing, in Windows prepares the terminal for ANSI output

**Comment by philbucher** [comment]
I added this to the committee TODO

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee  prefers to go for the proposal in #5643 as it would fit better in the `KratosLogger`. 

Does this PR introduce any feature missing in #5643 ?

**Comment by loumalouomega** [comment]
> Does this PR introduce any feature missing in #5643 ?

This is basically for Windows, #5643 is only Unix compatible

**Comment by loumalouomega** [comment]
I would recommend to check two PR, are changing different things

**Comment by philbucher** [comment]
the idea was to do the InitializeSTD somewhere in the logger

**Comment by loumalouomega** [comment]
> 
> 
> the idea was to do the InitializeSTD somewhere in the logger

Yep

**Comment by philbucher** [comment]
what is the status of this?

#6124 is merged

**Comment by loumalouomega** [comment]
I need to fix the windows compilation

File: CMakeLists.txt [hunk 3]
```diff
@@ -53,6 +53,12 @@ message("-- No KRATOS_BUILD_TESTING is defined. Setting to ON")
 SET (KRATOS_BUILD_TESTING ON)
 endif(NOT DEFINED KRATOS_BUILD_TESTING)
 
+# If no color policy not enable by default
+if(NOT DEFINED KRATOS_COLOR_STREAM)
```
Comment (philbucher) [inline CMakeLists.txt:60]: this is already there, don't remember the name

File: applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/penalty_frictional_mortar_criteria.h [hunk 4]
```diff
@@ -105,6 +105,9 @@ class PenaltyFrictionalMortarConvergenceCriteria
         const bool IODebug = false
         ) : BaseType(ComputeDynamicFactor, IODebug, PureSlip)
     {
+        // Initialize STD output
+        ColorUtilities::InitializeSTDOutput();
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/penalty_frictional_mortar_criteria.h:109]: didn't we say that this should be included in the logger?
Otherwise it is kinda unmaintainable

**Comment by philbucher** [comment]
II found that [catch2 solved this nicely](https://github.com/catchorg/Catch2/blob/master/include/internal/catch_console_colour.cpp)
To me looks like we can adapt this solution pretty straight forward
